From a7465bd970d4669df6b727b4038406f5bf043e1d Mon Sep 17 00:00:00 2001
From: Moh Aarif <mohammed.aarif@intel.com>
Date: Sun, 22 Mar 2020 19:35:08 +0530
Subject: [PATCH] UGW_SW-48995: This patch provides Ioctl interface to
 configure MAP-E mesh mode rules (FMRs - Forwarding Mapping Rules) from user
 space. This patch is ported from UGW-7.x. In UGW-7.x, Data Structures for map
 rules were taken from
 http://enog.jp/~masakazu/vyatta/map/iproute-2014-03-07.patch user space
 patch, and implementation also may be influenced with the content of this
 patch.

---
 include/uapi/linux/if_tunnel.h  |   1 +
 include/uapi/linux/ip6_tunnel.h |  38 ++++++
 net/ipv6/ip6_tunnel.c           | 277 +++++++++++++++++++++++++++++++++++++++-
 3 files changed, 312 insertions(+), 4 deletions(-)

diff --git a/include/uapi/linux/if_tunnel.h b/include/uapi/linux/if_tunnel.h
index ee42bf29c630..0afbc9c8eff8 100644
--- a/include/uapi/linux/if_tunnel.h
+++ b/include/uapi/linux/if_tunnel.h
@@ -20,6 +20,7 @@
 #define SIOCADD6RD      (SIOCDEVPRIVATE + 9)
 #define SIOCDEL6RD      (SIOCDEVPRIVATE + 10)
 #define SIOCCHG6RD      (SIOCDEVPRIVATE + 11)
+#define SIOCGETRULES    (SIOCDEVPRIVATE + 12)
 
 #define GRE_CSUM	__cpu_to_be16(0x8000)
 #define GRE_ROUTING	__cpu_to_be16(0x4000)
diff --git a/include/uapi/linux/ip6_tunnel.h b/include/uapi/linux/ip6_tunnel.h
index 48af63c9a48d..934fde56d1d8 100644
--- a/include/uapi/linux/ip6_tunnel.h
+++ b/include/uapi/linux/ip6_tunnel.h
@@ -48,4 +48,42 @@ struct ip6_tnl_parm2 {
 	__be32			o_key;
 };
 
+struct map_rule_parm {
+	struct in6_addr ipv6_prefix;
+	__u8            ipv6_prefix_length;
+	__be32          ipv4_prefix;
+	__u8            ipv4_prefix_length;
+	__u16           psid_prefix;
+	__u8            psid_prefix_length;
+	__u8            ea_length;
+	__u8            psid_offset;
+	__u8            forwarding_mode;
+	__u8            forwarding_rule;
+	struct in6_addr raddr;
+	struct in6_addr laddr;
+};
+
+struct map_pool_parm {
+	__be32          pool_prefix;
+	__u8            pool_prefix_length;
+};
+
+struct map_parm {
+	char                 name[IFNAMSIZ];
+	int                  tunnel_source;
+	struct in6_addr      br_address;
+	__u8                 br_address_length;
+	__u8                 role;
+	__u8                 default_forwarding_mode;
+	__u8                 default_forwarding_rule;
+	int                  ipv6_fragment_size;
+	__u8                 ipv4_fragment_inner;
+	__u8                 napt_always;
+	__u8                 napt_force_recycle;
+	unsigned long        rule_num;
+	unsigned long        pool_num;
+	struct map_rule_parm rule[0];
+	struct map_pool_parm pool[0];
+};
+
 #endif
diff --git a/net/ipv6/ip6_tunnel.c b/net/ipv6/ip6_tunnel.c
index 9671aa81d880..c17b5cd87ac6 100644
--- a/net/ipv6/ip6_tunnel.c
+++ b/net/ipv6/ip6_tunnel.c
@@ -43,7 +43,7 @@
 #include <linux/slab.h>
 #include <linux/hash.h>
 #include <linux/etherdevice.h>
-
+#include <linux/inet.h>
 #include <asm/uaccess.h>
 #include <linux/atomic.h>
 
@@ -84,6 +84,12 @@ static u32 HASH(const struct in6_addr *addr)
 static int ip6_tnl_dev_init(struct net_device *dev);
 static void ip6_tnl_dev_setup(struct net_device *dev);
 static struct rtnl_link_ops ip6_link_ops __read_mostly;
+static int ipmap_parm_from_user(struct map_rule_parm *p_rule,
+	struct __ip6_tnl_parm *p2, struct net *net, struct ip6_tnl **tl);
+static void ipmap_parm_to_user(struct __ip6_tnl_fmr *fmrs,
+	struct map_rule_parm *p_rule);
+static int del_fmr(struct map_rule_parm *p_rule, struct net *net);
+static int if_map_parm_unique(struct ip6_tnl *t, struct map_rule_parm *p_rule);
 
 static int ip6_tnl_net_id __read_mostly;
 struct ip6_tnl_net {
@@ -1387,6 +1393,7 @@ ip4ip6_tnl_xmit(struct sk_buff *skb, struct net_device *dev)
 {
 	struct ip6_tnl *t = netdev_priv(dev);
 	const struct iphdr  *iph;
+	struct __ip6_tnl_fmr *fmr;
 	int encap_limit = -1;
 	struct flowi6 fl6;
 	__u8 dsfield;
@@ -1436,6 +1443,19 @@ ip4ip6_tnl_xmit(struct sk_buff *skb, struct net_device *dev)
 
 	dsfield = INET_ECN_encapsulate(dsfield, ipv4_get_dsfield(iph));
 
+	/* try to find matching FMR (Forwarding Mapping Rules) */
+	for (fmr = t->parms.fmrs; fmr; fmr = fmr->next) {
+		unsigned mshift = 32 - fmr->ip4_prefix_len;
+		if (ntohl(fmr->ip4_prefix.s_addr) >> mshift ==
+				ntohl(ip_hdr(skb)->daddr) >> mshift)
+			break;
+	}
+
+	/* change dstaddr according to FMR */
+	if (fmr)
+		ip4ip6_fmr_calc(&fl6.daddr, ip_hdr(skb),
+			skb_tail_pointer(skb), fmr, true);
+
 	if (iptunnel_handle_offloads(skb, SKB_GSO_IPXIP6))
 		return -1;
 
@@ -1534,7 +1554,8 @@ ip6ip6_tnl_xmit(struct sk_buff *skb, struct net_device *dev)
 
 	/* change dstaddr according to FMR */
 	if (fmr)
-		ip4ip6_fmr_calc(&fl6.daddr, ip_hdr(skb), skb_tail_pointer(skb), fmr, true);
+		ip4ip6_fmr_calc(&fl6.daddr, ip_hdr(skb),
+			skb_tail_pointer(skb), fmr, true);
 
 	if (iptunnel_handle_offloads(skb, SKB_GSO_IPXIP6))
 		return -1;
@@ -1725,6 +1746,147 @@ ip6_tnl_parm_to_user(struct ip6_tnl_parm *u, const struct __ip6_tnl_parm *p)
 	memcpy(u->name, p->name, sizeof(u->name));
 }
 
+static int if_map_parm_unique(struct ip6_tnl *t, struct map_rule_parm *p_rule)
+{
+	struct __ip6_tnl_fmr *fmr = NULL;
+	fmr = t->parms.fmrs;
+	while (fmr) {
+		/*Compare v6 prefix, v4 prefix and ea-length*/
+		if ((ipv6_prefix_equal(&p_rule->ipv6_prefix, &fmr->ip6_prefix,
+			p_rule->ipv6_prefix_length))
+			&& ((ntohl(fmr->ip4_prefix.s_addr)
+			>> (32 - fmr->ip4_prefix_len))
+			== ((ntohl(p_rule->ipv4_prefix))
+			>> (32 - p_rule->ipv4_prefix_length)))
+			&& (fmr->ea_len == p_rule->ea_length)) {
+			return 1;
+		}
+		fmr = fmr->next;
+	}
+	return 0;
+}
+
+static void ipmap_parm_to_user(struct __ip6_tnl_fmr *fmrs,
+	struct map_rule_parm *p_rule)
+{
+	p_rule->ipv6_prefix_length = fmrs->ip6_prefix_len;
+	p_rule->ipv4_prefix_length = fmrs->ip4_prefix_len;
+	p_rule->ea_length = fmrs->ea_len;
+	p_rule->psid_offset = 6; /* fix 6 bits for offset */
+	memcpy(&p_rule->ipv6_prefix, &fmrs->ip6_prefix,
+		sizeof(fmrs->ip6_prefix));
+	memcpy(&p_rule->ipv4_prefix, &fmrs->ip4_prefix,
+		sizeof(fmrs->ip4_prefix));
+}
+
+static int ipmap_parm_from_user(struct map_rule_parm *p_rule,
+	struct __ip6_tnl_parm *p2, struct net *net, struct ip6_tnl **tnl)
+{
+	struct __ip6_tnl_fmr *t_parm = NULL, *fmr_node = NULL;
+	struct ip6_tnl *t = NULL;
+	p2->laddr = p_rule->laddr;
+	p2->raddr = p_rule->raddr;
+
+	t = ip6_tnl_locate(net, p2, 0);
+	if (!IS_ERR(t)) {
+		if (if_map_parm_unique(t, p_rule))
+			return -EINVAL;
+		p2->fmrs = kzalloc(sizeof(struct __ip6_tnl_fmr), GFP_KERNEL);
+		if (!p2->fmrs)
+			return -ENOMEM;
+		p2->fmrs->next = NULL;
+		p2->fmrs->ip6_prefix_len = p_rule->ipv6_prefix_length;
+		p2->fmrs->ip4_prefix_len = p_rule->ipv4_prefix_length;
+		p2->fmrs->ea_len = p_rule->ea_length;
+		p2->fmrs->offset = p_rule->psid_offset;
+		memcpy(&p2->fmrs->ip6_prefix, &p_rule->ipv6_prefix,
+			sizeof(p2->fmrs->ip6_prefix));
+		memcpy(&p2->fmrs->ip4_prefix, &p_rule->ipv4_prefix,
+			sizeof(p2->fmrs->ip4_prefix));
+		p2->laddr = t->parms.laddr;
+		p2->raddr = t->parms.raddr;
+		p2->flags = t->parms.flags;
+		p2->hop_limit = t->parms.hop_limit;
+		p2->encap_limit = t->parms.encap_limit;
+		p2->flowinfo = t->parms.flowinfo;
+		p2->link = t->parms.link;
+		p2->proto = t->parms.proto;
+		memcpy(p2->name, t->parms.name, sizeof(t->parms.name));
+		t_parm = t->parms.fmrs;
+
+		/* Copy fmr link list from the tunnel parm*/
+		while (t_parm != NULL) {
+			fmr_node = kzalloc(sizeof(struct __ip6_tnl_fmr),
+				GFP_KERNEL);
+			if (!fmr_node)
+				return -ENOMEM;
+			memcpy(fmr_node, t_parm, sizeof(struct __ip6_tnl_fmr));
+			fmr_node->next = p2->fmrs;
+			p2->fmrs = fmr_node;
+			t_parm = t_parm->next;
+		}
+	} else {
+		return -1;
+	}
+	*tnl = t;
+	return 0;
+}
+
+static int del_fmr(struct map_rule_parm *p_rule, struct net *net)
+{
+	int fmr_del = -ENOENT;
+	struct ip6_tnl *t = NULL;
+	struct __ip6_tnl_parm p2;
+	struct __ip6_tnl_fmr *prev_fmrs = NULL, *fmrs = NULL, *temp = NULL;
+	memcpy(&p2.laddr, &p_rule->laddr, sizeof(p2.laddr));
+	memcpy(&p2.raddr, &p_rule->raddr, sizeof(p2.raddr));
+
+	/* Locate the tunnel from the given remote and local address*/
+	t = ip6_tnl_locate(net, &p2, 0);
+	if (!IS_ERR(t))
+		return -ENOENT;
+	fmrs = t->parms.fmrs;
+	temp = t->parms.fmrs;
+	prev_fmrs = t->parms.fmrs;
+
+	/* Find the node by matching the Param "IPv6 Prefix,
+		IPv4 Prefix & EA LENGTH" */
+	if (t->parms.fmrs) {
+		if (((ipv6_prefix_equal(&p_rule->ipv6_prefix,
+			&fmrs->ip6_prefix, p_rule->ipv6_prefix_length)))
+			&& ((ntohl(fmrs->ip4_prefix.s_addr)
+			>> (32 - fmrs->ip4_prefix_len))
+			== ((ntohl(p_rule->ipv4_prefix))
+			>> (32 - p_rule->ipv4_prefix_length)))
+			&& (fmrs->ea_len == p_rule->ea_length)) {
+			fmr_del = 1;
+			t->parms.fmrs = fmrs->next;
+			kfree(fmrs);
+		} else {
+			while (fmrs != NULL) {
+				if (((ipv6_prefix_equal(&p_rule->ipv6_prefix,
+					&fmrs->ip6_prefix,
+					p_rule->ipv6_prefix_length)))
+					&& ((ntohl(fmrs->ip4_prefix.s_addr)
+					>> (32 - fmrs->ip4_prefix_len))
+					== ((ntohl(p_rule->ipv4_prefix))
+					>> (32 - p_rule->ipv4_prefix_length)))
+					&& (fmrs->ea_len
+					== p_rule->ea_length)) {
+					prev_fmrs->next = fmrs->next;
+					kfree(fmrs);
+					fmr_del = 1;
+					break;
+				}
+				prev_fmrs = fmrs;
+				fmrs = fmrs->next;
+			}
+		}
+	}
+
+	return fmr_del;
+}
+
 /**
  * ip6_tnl_ioctl - configure ipv6 tunnels from userspace
  *   @dev: virtual device associated with tunnel
@@ -1756,12 +1918,15 @@ ip6_tnl_parm_to_user(struct ip6_tnl_parm *u, const struct __ip6_tnl_parm *p)
 static int
 ip6_tnl_ioctl(struct net_device *dev, struct ifreq *ifr, int cmd)
 {
-	int err = 0;
+	int err = 0, fmrs_count = 0, i;
 	struct ip6_tnl_parm p;
-	struct __ip6_tnl_parm p1;
+	struct __ip6_tnl_parm p1, p2;
 	struct ip6_tnl *t = netdev_priv(dev);
+	struct __ip6_tnl_fmr *temp = NULL;
 	struct net *net = t->net;
 	struct ip6_tnl_net *ip6n = net_generic(net, ip6_tnl_net_id);
+	struct map_parm *pptr, pParm;
+	struct map_rule_parm *p_rule;
 
 	memset(&p1, 0, sizeof(p1));
 
@@ -1784,6 +1949,7 @@ ip6_tnl_ioctl(struct net_device *dev, struct ifreq *ifr, int cmd)
 			err = -EFAULT;
 		}
 		break;
+
 	case SIOCADDTUNNEL:
 	case SIOCCHGTUNNEL:
 		err = -EPERM;
@@ -1843,6 +2009,109 @@ ip6_tnl_ioctl(struct net_device *dev, struct ifreq *ifr, int cmd)
 		err = 0;
 		unregister_netdevice(dev);
 		break;
+	case SIOCGETRULES:
+	case SIOCGETPRL:
+		if (t == NULL)
+			t = netdev_priv(dev);
+		if (t != NULL) {
+			if (cmd == SIOCGETRULES) {
+				if (t->parms.fmrs != NULL) {
+					temp = t->parms.fmrs;
+					while (temp != NULL) {
+						temp = temp->next;
+						fmrs_count++;
+					}
+					if (!fmrs_count)
+						err = ENOENT;
+					pParm.rule_num = fmrs_count;
+					if (copy_to_user
+						(ifr->ifr_ifru.ifru_data,
+						&pParm,
+						sizeof(struct map_parm)))
+						err = -EFAULT;
+				} else {
+					err = ENOENT;
+				}
+			}
+			if (cmd == SIOCGETPRL) {
+				temp = t->parms.fmrs;
+				while (temp != NULL) {
+					temp = temp->next;
+					fmrs_count++;
+				}
+				pptr = kzalloc(sizeof(struct map_parm)
+					+ (sizeof(struct map_rule_parm)
+					* fmrs_count), GFP_KERNEL);
+				if (!pptr)
+					break;
+				temp = t->parms.fmrs;
+				for (i = 0; i < fmrs_count; i++) {
+					if (temp != NULL) {
+						ipmap_parm_to_user(temp,
+							&pptr->rule[i]);
+						temp = temp->next;
+					}
+				}
+				pptr->rule_num = fmrs_count;
+				if (copy_to_user(ifr->ifr_ifru.ifru_data, pptr,
+					sizeof(struct map_parm)
+					+ (sizeof(struct map_rule_parm)
+					* fmrs_count)))
+					err = -EFAULT;
+				if (pptr != NULL)
+					kfree(pptr);
+			}
+		} else {
+			err = -EINVAL;
+		}
+		break;
+	case SIOCADDPRL:
+		err = -EPERM;
+		if (!ns_capable(net->user_ns, CAP_NET_ADMIN))
+			break;
+		pptr = kzalloc(sizeof(struct map_parm)
+			+ sizeof(struct map_rule_parm), GFP_KERNEL);
+		if (!pptr)
+			break;
+		memset(&p2, 0, sizeof(p2));
+		if (copy_from_user(pptr, ifr->ifr_ifru.ifru_data,
+			sizeof(struct map_parm)
+			+ sizeof(struct map_rule_parm)))
+			break;
+		p_rule = &pptr->rule[0];
+
+		if (ipmap_parm_from_user(p_rule, &p2, net, &t) == 0) {
+			err = ip6_tnl_update(t, &p2);
+			ipmap_parm_to_user(t->parms.fmrs, p_rule);
+			if (copy_to_user(ifr->ifr_ifru.ifru_data,
+				pptr, sizeof(struct map_parm)
+				+ sizeof(struct map_rule_parm)))
+				err = -EFAULT;
+
+		} else {
+			err = -ENOENT;
+		}
+		kfree(pptr);
+		break;
+	case SIOCDELPRL:
+		err = -EPERM;
+		if (!ns_capable(net->user_ns, CAP_NET_ADMIN))
+			break;
+		pptr = kzalloc(sizeof(struct map_parm)
+			+ sizeof(struct map_rule_parm), GFP_KERNEL);
+		if (!pptr)
+			break;
+		memset(&p2, 0, sizeof(p2));
+		if (copy_from_user(pptr, ifr->ifr_ifru.ifru_data,
+			sizeof(struct map_parm)
+			+ sizeof(struct map_rule_parm)))
+			break;
+		p_rule = &pptr->rule[0];
+		err = 0;
+		if (!del_fmr(p_rule, net))
+			err = -ENOENT;
+		kfree(pptr);
+		break;
 	default:
 		err = -EINVAL;
 	}
